### PR TITLE
I/O redirection part 2: redirecting stdin + large refactoring + exe/sort.js

### PR DESCRIPTION
**Pull request by ussuri**
@rpaquay 

Again, sorry for the large PR :)

Some notes:

* Added a notion of closing for streams/buffers/stdio:
 * This allows a producer or consumer of a stream/buffer to indicate that there is no more data to write (for writable), or that it's no longer interested in the data (for readable), so that the consumer/producer on the other end can stop writing/reading.
 * This is critical for filter-like execs.
 * The notion of closing right now is just a function call and a corresponding event: it doesn't yet go as far tracking the state of the stream/buffer at all times; and the event can theoretically be fired multiple times.
 * I think that to take it further, stdio, streams and buffers should be made proper `Ephemerals`.
 * I haven't completely solved the chicken and egg problem where, for example, buffers create streams: it the owned stream gets closed by the client, should the owning buffer close as well? What about the opposite? I've implemented some logic, but I'm not yet 100% sure it's right. We could iterate on this.
* After the first redirection PR, Wash.prototype.dispatch and ExecuteContext.prototype.setCallee ended up created two levels of NestedStdio: after adding the notion of closing for streams, that created problems with various parties listening to `onClose` on one or the other of this two nested `NestedStdio`s.
* Saving both `stdio` and its individual streams on `ExecuteContext` creating an aliasing problem: nothing prevented the client from directly resetting one and getting the whole set out of sync (happened to me during debugging). I'd prefer to keep just the `stdio` member; however, there we so much usage of `stdin` etc. in the code, that I decided to do the opposite for now.
* Added exe/sort.js (used it as a testing tool).

I'll add more comments directly in the code changes tomorrow.

### CODE REVIEW

File: lib/axiom/fs/nested_stdio.js [hunk 1]
```diff
@@ -15,22 +15,21 @@
...
+  // handlers added are scoped to the nested stdio lifetime.
   var stdin = new ReadableStreamForwarder(parentStdio.stdin);
   var signal = new ReadableStreamForwarder(parentStdio.signal);
   var ttyin = new ReadableStreamForwarder(parentStdio.ttyin);
-  // Note: stdout, stderr and ttyout don't need to be wrapped as they don't
-  // expose events.
-  var stdout = parentStdio.stdout;
-  var stderr = parentStdio.stderr;
-  var ttyout = parentStdio.ttyout;
+  var stdout = new WritableStreamForwarder(parentStdio.stdout);
```
Comment (ussuri) [inline lib/axiom/fs/nested_stdio.js]: Output streams now also have an event to fire (onClose), hence the need to wrap them as well.

File: lib/axiom/fs/stream/memory_stream_buffer.js [hunk 2]
```diff
@@ -14,9 +14,9 @@
 
 import AxiomError from 'axiom/core/error';
 import AxiomEvent from 'axiom/core/event';
-import Queue from 'axiom/fs/stream/queue';
 import ReadableMemoryStream from 'axiom/fs/stream/readable_memory_stream';
 import WritableMemoryStream from 'axiom/fs/stream/writable_memory_stream';
+import StreamBuffer from 'axiom/fs/stream/stream_buffer';
```
Comment (ussuri) [inline lib/axiom/fs/stream/memory_stream_buffer.js]: Most of the MemoryStreamBuffer has migrated to a new generic StreamBuffer.

File: lib/axiom/fs/stream/writable_memory_stream.js [hunk 3]
```diff
@@ -13,36 +13,30 @@
...
 /** @typedef function():void */
 var EventCallback;
 
 /**
- * Event value + associated callback
- * @constructor
- * @param {!*} value
- * @param {!EventCallback} callback
- */
-var EventWithCallback = function(value, callback) {
```
Comment (ussuri) [inline lib/axiom/fs/stream/writable_memory_stream.js]: This wasn't used

File: lib/wash/exe/wash.js [hunk 4]
```diff
@@ -532,33 +533,53 @@ Wash.prototype.readEvalPrintLoop = function() {
...
+
+/**
+ * Matches a command line with possible stdout redirection to a file:
+ *
+ *    command [arg1 arg2 ...] [< some_file] [> some_file]
+ *
+ * @private
+ * @const
+ */
+Wash.CMD_WITH_IO_REDIRECT_RE_ = /^([^<>]+?)(?:\s+([^<>]+?))?\s*([<>].+)?$/;
```
Comment (ussuri) [inline lib/wash/exe/wash.js]: Yes, regexps :) If you feel uneasy about them, we can iterate, but right now I'm out of steam with this PR (in addition to not being a fan of the manual for-if-else-if-else-if kind of parsers).

File: lib/axiom/core/event.js [hunk 5]
```diff
@@ -45,6 +45,74 @@ export {AxiomEvent};
...
+ * (with opt_fireOnce off) one of the sub-events fire.
+ *
+ * @param {!Array<!AxiomEvent>} events
+ * @param {boolean=} opt_fireOnce
+ * @return {!AxiomEvent}
+ */
+AxiomEvent.any = function(events, opt_fireOnce) {
+  var fireOnce = opt_fireOnce || false;
+  var event = new AxiomEvent();
+
```
Comment (rpaquay) [inline lib/axiom/core/event.js]: extra whitespace

File: lib/axiom/core/event.js [hunk 6]
```diff
@@ -45,6 +45,74 @@ export {AxiomEvent};
 export default AxiomEvent;
 
 /**
+ * The returned meta-event will fire as soon as (with opt_fireOnce) or whenever
```
Comment (rpaquay) [inline lib/axiom/core/event.js]: I have a hard time understanding this comment, because it does not start with a verb. Maybe something along the lines of:
```
Create an event that fires whenever any of the events passed as parameter fires.
When opt_fireOnce is true, the event fires only once, otherwise the event fires every time.
```

File: lib/axiom/core/event.js [hunk 7]
```diff
@@ -45,6 +45,74 @@ export {AxiomEvent};
...
+  var firedEvents = [];
+  var listeners = [];
+  
+  var listener = function(eventIdx, var_args) {
+    firedEvents[eventIdx] = true;
+    if (firedEvents.every(function(elt) { return elt === true; })) {
+      var args = Array.prototype.slice.call(arguments, 1);
+      event.fire.apply(event, args);
+      if (!fireOnce) {
+        for (var i = 0; i < events.length; ++i) {
```
Comment (rpaquay) [inline lib/axiom/core/event.js]: s/events/firedEvents

File: lib/axiom/core/event.js [hunk 8]
```diff
@@ -45,6 +45,74 @@ export {AxiomEvent};
...
 
 /**
+ * The returned meta-event will fire as soon as (with opt_fireOnce) or whenever
+ * (with opt_fireOnce off) one of the sub-events fire.
+ *
+ * @param {!Array<!AxiomEvent>} events
+ * @param {boolean=} opt_fireOnce
+ * @return {!AxiomEvent}
+ */
+AxiomEvent.any = function(events, opt_fireOnce) {
```
Comment (rpaquay) [inline lib/axiom/core/event.js]: It would nice to add unit tests for this.

File: lib/axiom/core/event.js [hunk 9]
```diff
@@ -45,6 +45,74 @@ export {AxiomEvent};
...
+/**
+ * The returned meta-event will fire as soon as all of the sub-events have fired
+ * at least once. After that, the process either stops (with opt_fireOnce on)
+ * or resets (with opt_fireOnce on).
+ * 
+ * @param {!Array<!AxiomEvent>} events
+ * @param {boolean=} opt_fireOnce
+ * @return {!AxiomEvent}
+ */
+AxiomEvent.all = function(events, opt_fireOnce) {
```
Comment (rpaquay) [inline lib/axiom/core/event.js]: It would nice to add unit tests for this.

File: lib/axiom/fs/base/execute_context.js [hunk 10]
```diff
@@ -213,6 +205,27 @@ ExecuteContext.prototype.setStdio = function(stdio) {
 };
 
 /**
+ * @return {!Stdio}
+ */
+ExecuteContext.prototype.getStdio = function() {
```
Comment (rpaquay) [inline lib/axiom/fs/base/execute_context.js]: I don't understand why this is needed. If you had left this.stdio = stdio in the constructor, this function would not be needed, right?

File: lib/axiom/fs/base/execute_context.js [hunk 11]
```diff
@@ -94,8 +94,6 @@ export var ExecuteContext = function(fileSystem, stdio, path, args) {
    */
   this.callee = null;
 
-  /** @type {!Stdio} */
-  this.stdio = stdio;
```
Comment (rpaquay) [inline lib/axiom/fs/base/execute_context.js]: Even though this does not break anything right now, it would be better to leave this as a member. stdio is meant to be a container of any number of streams (see comment in `stdio.js`: Additional streams can be added as named properties.). The stdin, etc. members are just there for convenience shortcuts.

See my comment in `getNestedStdio` on why it would be better to leave stdio as a member.

File: lib/axiom/fs/base/execute_context.js [hunk 12]
```diff
@@ -213,6 +205,27 @@ ExecuteContext.prototype.setStdio = function(stdio) {
...
+      this.ttyin,
+      this.ttyout
+  );
+};
+
+/**
+ * @return {!NestedStdio}
+ */
+ExecuteContext.prototype.createNestedStdio = function() {
+  return new NestedStdio(this.getStdio());
```
Comment (rpaquay) [inline lib/axiom/fs/base/execute_context.js]: This does not break anything right now, but it will soon, because stdio is meant to be a container of any number of streams. So, the NestedStdio constructor needs to have the original stdio (with all the additional streams) to do its work of creating the "wrapper" streams for all the streams (not just the predefined ones).

File: lib/axiom/fs/base/execute_context.js [hunk 13]
```diff
@@ -362,7 +375,7 @@ ExecuteContext.prototype.delEnv = function(name) {
  */
 ExecuteContext.prototype.createExecuteContext = function(path, arg) {
   return this.fileSystemManager.createExecuteContext(
-      path, this.stdio, arg).then(
+      path, this.getStdio(), arg).then(
```
Comment (rpaquay) [inline lib/axiom/fs/base/execute_context.js]: this.stdio was correct, right?

File: lib/axiom/fs/nested_stdio.js (hunk 1 above)
Comment (rpaquay) [inline lib/axiom/fs/nested_stdio.js]: Ack.

File: lib/axiom/fs/stdio.js [hunk 14]
```diff
@@ -47,3 +47,15 @@ export var Stdio = function(stdin, stdout, stderr, signal, ttyin, ttyout) {
 };
 
 export default Stdio;
+
+/**
+ * @return {void}
+ */
+Stdio.prototype.close = function() {
+  this.stdin.close();
```
Comment (rpaquay) [inline lib/axiom/fs/stdio.js]: (Sorry I am reading this sequentially, so my comment here will probably be duplicated somewhere else):
ReadableStream should not have a "close" method. We are following the node.js stream design (see https://nodejs.org/api/stream.html). You can have a "close" or "end" event, but the *consumer* of a readable stream should not be able to close it.

File: lib/axiom/fs/stdio.js [hunk 15]
```diff
@@ -47,3 +47,15 @@ export var Stdio = function(stdin, stdout, stderr, signal, ttyin, ttyout) {
...
+  this.stdout.close();
```
Comment (rpaquay) [inline lib/axiom/fs/stdio.js]: For WritableStream, it should be a method named `end` (see https://nodejs.org/api/stream.html). I think we can leave the `end` method without any argument for now, unlike node.js streams.

File: lib/axiom/fs/stream/memory_stream_buffer.js (hunk 2 above)
Comment (rpaquay) [inline lib/axiom/fs/stream/memory_stream_buffer.js]: Ack

File: lib/axiom/fs/stream/memory_stream_buffer.js [hunk 16]
```diff
@@ -24,114 +24,33 @@ var ReadableStream;
...
-  /** @type {!AxiomEvent} */
-  this.onReadable = new AxiomEvent();
+  StreamBuffer.call(this);
+
   /** @const @type {!ReadableStream} */
   this.readableStream = new ReadableMemoryStream(this);
   /** @const @type {!WritableStream} */
   this.writableStream = new WritableMemoryStream(this);
+  
+  this.onClose = AxiomEvent.all([
```
Comment (rpaquay) [inline lib/axiom/fs/stream/memory_stream_buffer.js]: I think I see what you are doing here, but given my previous comments, I think it should be something along the lines of
* add an `end` method
* when `end` is called, flush all write events, then call the `onEnd' event on the readable stream and the `onFinish` event on the writable stream.

File: lib/axiom/fs/stream/memory_stream_buffer.js [hunk 17]
```diff
@@ -24,114 +24,33 @@ var ReadableStream;
...
- */
-MemoryStreamBuffer.prototype.read_ = function() {
-  var item = this.events_.dequeue();
-  if (item instanceof EventWithCallback) {
-    item.callback();
-    return item.value;
-  } else {
-    return item;
-  }
+MemoryStreamBuffer.prototype.close = function() {
```
Comment (rpaquay) [inline lib/axiom/fs/stream/memory_stream_buffer.js]: Oh I see... So, see my comments above: this should be named `end`, and should flush all writes before firing `onEnd` and `onFinish`.

File: lib/axiom/fs/stream/readable_file_stream_buffer.js [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+ */
+export var ReadableFileStreamBuffer = function(fileSystem, filePath, openMode) {
+  StreamBuffer.call(this);
+
+  /** @const @type {!ReadableStream} */
+  this.readableStream = new ReadableMemoryStream(this);
+  
+  this.onClose = new AxiomEvent();
+
+  fileSystem.readFile(filePath, DataType.UTF8String, openMode)
```
Comment (rpaquay) [inline lib/axiom/fs/stream/readable_file_stream_buffer.js]: What about errors?

Thinking about this, and still looking at the way node.js does things, I think we don't need this class. I think we should just create a `createReadableStream` method on FileSystem: The method would be the code here, except it would buffer all lines into a `MemoryBufferStream` for now. For FileSystem's that have "real" streams available under the hood (e.g. node.js), they can override `createReableableStream` and wrap their stream implementation to expose a `ReadableStream` interface. This would allow reading very large files from the local file system without having to load everything in memory.

File: lib/axiom/fs/stream/readable_memory_stream.js [hunk 19]
```diff
@@ -63,3 +65,11 @@ ReadableMemoryStream.prototype.resume = function() {
 ReadableMemoryStream.prototype.read = function() {
   return this.buffer_.read();
 };
+
+/**
+ * @override
+ * @return {void}
+ */
+ReadableMemoryStream.prototype.close = function() {
```
Comment (rpaquay) [inline lib/axiom/fs/stream/readable_memory_stream.js]: There should be no close method, as explained earlier.

File: lib/axiom/fs/stream/readable_stream.js [hunk 20]
```diff
@@ -39,6 +39,12 @@ export var ReadableStream = function() {
    * @type {!AxiomEvent}
    */
   this.onReadable = new AxiomEvent();
+  /**
+   * Fires when the stream gets closed.
+   *
+   * @type {!AxiomEvent}
+   */
+  this.onClose = new AxiomEvent();
```
Comment (rpaquay) [inline lib/axiom/fs/stream/readable_stream.js]: s/onClose/onEnd/

File: lib/axiom/fs/stream/readable_stream.js [hunk 21]
```diff
@@ -63,3 +69,11 @@ ReadableStream.prototype.resume = function() {};
...
  */
 ReadableStream.prototype.read = function() {};
+
+/**
+ * Close the stream when there is no more data to stream (e.g. an EOF is
+ * reached in the underlying file).
+ *
+ * @return {void}
+ */
+ReadableStream.prototype.close = function() {};
```
Comment (rpaquay) [inline lib/axiom/fs/stream/readable_stream.js]: remove

File: lib/axiom/fs/stream/readable_stream_forwarder.js [hunk 22]
```diff
@@ -28,10 +28,13 @@ export var ReadableStreamForwarder = function(stream) {
   this.onData = new AxiomEvent();
   /** type {!AxiomEvent} */
   this.onReadable = new AxiomEvent();
+  /** type {!AxiomEvent} */
+  this.onClose = new AxiomEvent();
```
Comment (rpaquay) [inline lib/axiom/fs/stream/readable_stream_forwarder.js]: s/onClose/onEnd

File: lib/axiom/fs/stream/writable_file_stream_buffer.js [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+ */
+export var WritableFileStreamBuffer = function(fileSystem, filePath, fileMode) {
+  StreamBuffer.call(this);
+
+  /** @const @type {!WritableStream} */
+  this.writableStream = new WritableMemoryStream(this);
+
+  this.onClose = new AxiomEvent();
+
+  fileSystem.createOpenContext(filePath, fileMode)
```
Comment (rpaquay) [inline lib/axiom/fs/stream/writable_file_stream_buffer.js]: Same comment as `ReadableFileStreamBuffer`: I don't think we need this class, we can just have a `createWritableStream` on `FileSystem` with a default implementation.

File: lib/axiom/fs/stream/writable_memory_stream.js (hunk 3 above)
Comment (rpaquay) [inline lib/axiom/fs/stream/writable_memory_stream.js]: Ack

File: lib/axiom/fs/stream/writable_stream.js [hunk 24]
```diff
@@ -12,15 +12,24 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import AxiomEvent from 'axiom/core/event';
+
 /** @typedef function():void */
-var WriteCallback;
+export var WriteCallback;
```
Comment (rpaquay) [inline lib/axiom/fs/stream/writable_stream.js]: Why is this needed?

File: lib/wash/exe/sort.js [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+      'and prints the result to stdout.'
+    ].join('\r\n') + '\r\n');
+    cx.closeOk();
+    return;
+  }
+
+  var lines = [];
+
+  cx.stdin.onReadable.addListener(function() {
+    lines.push(cx.stdin.read());
```
Comment (rpaquay) [inline lib/wash/exe/sort.js]: Technically, this should be a while loop, for as long `read` returns a value, because `onReadable` might be fired less frequently than read data is available.

That said, maybe a better way would be to use `onData` event, and explicitly call `stdin.resume()` after hooking up the event.

File: lib/axiom/core/event.js (hunk 6 above)
Comment (ussuri) [inline lib/axiom/core/event.js]: Sounds good. Done.

File: lib/axiom/core/event.js (hunk 5 above)
Comment (ussuri) [inline lib/axiom/core/event.js]: done

File: lib/axiom/core/event.js (hunk 8 above)
Comment (ussuri) [inline lib/axiom/core/event.js]: Absolutely. Will do.

File: lib/axiom/core/event.js (hunk 9 above)
Comment (ussuri) [inline lib/axiom/core/event.js]: Will do.

File: lib/axiom/core/event.js (hunk 7 above)
Comment (ussuri) [inline lib/axiom/core/event.js]: The stress here is on `events` being the source of truth: `fireEvents` is initialized from scratch, in the same way it's initially done in the setup code further below. I've just made it more explicit by  setting `firedEvents` to `[]` before the loop.

File: lib/axiom/fs/base/execute_context.js (hunk 10 above)
Comment (ussuri) [inline lib/axiom/fs/base/execute_context.js]: This is needed because:
```
cx.setStdio(stdio1);
...  
cx.stdin = stdin2;
...
cx.stdio.stdin.addListener(); // attaches to stdio1.stdin
cx.stdin.addListener(); // attaches to stdin2
```
I would get rid of `ExecuteContext.getStdio()` as well, but some client code still wants to apply group actions to the whole virtual `ExecuteContext.stdio` at once.

File: lib/axiom/fs/base/execute_context.js (hunk 13 above)
Comment (ussuri) [inline lib/axiom/fs/base/execute_context.js]: Yes, but `this.stdio` was replaced with `this.getStdio()` in the new code.

File: lib/axiom/fs/stdio.js (hunk 14 above)
Comment (ussuri) [inline lib/axiom/fs/stdio.js]: Sounds reasonable

File: lib/axiom/fs/base/execute_context.js (hunk 10 above)
Comment (ussuri) [inline lib/axiom/fs/base/execute_context.js]: An alternative could be to leave just `this.stdio` and get rid of `this.stdin` etc. But the client code was using the latter in 99% of the cases, so I went with the minimal change for now.

File: lib/wash/exe/sort.js (hunk 25 above)
Comment (ussuri) [inline lib/wash/exe/sort.js]: done

File: lib/axiom/fs/stream/writable_stream.js (hunk 24 above)
Comment (ussuri) [inline lib/axiom/fs/stream/writable_stream.js]: I thought it was needed because it was used in a Closure comment in `writable_stream_forwarder.js`, but just checked, and the error I was getting was due to something else. Reverted.

**Comment by ussuri** [comment]
#226 replaces this PR
